### ISSUE TITLE
BatchNormalization after Flatten gives "Duplicate binding" error

### ISSUE THREAD

**Issue by HansPinckaers** [plaidml/plaidml#57]
Hi,

Thanks for all the work on PlaidML! 

I'm running into an issue using a BatchNorm layer after flatten. 

Sample code:
```
import numpy as np

# Install the plaidml backend
import plaidml.keras
plaidml.keras.install_backend()

import keras
from keras.layers.normalization import BatchNormalization
from keras.layers import Input, Dense, Flatten
from keras.models import Model

# This returns a tensor
inputs = Input(shape=(2,2))

# Removing the Flatten makes the model run,
# Changing the order of BatchNorm and Flatten also runs
x = Flatten()(inputs)
x = BatchNormalization()(x)
predictions = Dense(2, activation='softmax')(x)

model = Model(inputs=inputs, outputs=predictions)
model.summary()

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

x_train = np.array([[[1,1],[1,1]]])

y = model.predict(x=x_train, batch_size=1)
model.fit(x_train, np.array([[0., 1.]]), batch_size=1, epochs=1, verbose=1)
```

Error message:
```
INFO:plaidml:b'Opening device "amd_radeon_r9_m370x_compute_engine.0"'
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         (None, 2, 2)              0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 4)                 0         
_________________________________________________________________
batch_normalization_1 (Batch (None, 4)                 16        
_________________________________________________________________
dense_1 (Dense)              (None, 2)                 10        
=================================================================
Total params: 26
Trainable params: 18
Non-trainable params: 8
_________________________________________________________________
Epoch 1/1
---------------------------------------------------------------------------
Unknown                                   Traceback (most recent call last)
<ipython-input-1-bed6b5bd5b5f> in <module>()
     27 
     28 y = model.predict(x=x_train, batch_size=1)
---> 29 model.fit(x_train, np.array([[0., 1.]]), batch_size=1, epochs=1, verbose=1)

~/test/lib/python3.5/site-packages/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, **kwargs)
   1596                               initial_epoch=initial_epoch,
   1597                               steps_per_epoch=steps_per_epoch,
-> 1598                               validation_steps=validation_steps)
   1599 
   1600     def evaluate(self, x, y,

~/test/lib/python3.5/site-packages/keras/engine/training.py in _fit_loop(self, f, ins, out_labels, batch_size, epochs, verbose, callbacks, val_f, val_ins, shuffle, callback_metrics, initial_epoch, steps_per_epoch, validation_steps)
   1181                     batch_logs['size'] = len(batch_ids)
   1182                     callbacks.on_batch_begin(batch_index, batch_logs)
-> 1183                     outs = f(ins_batch)
   1184                     if not isinstance(outs, list):
   1185                         outs = [outs]

~/test/lib/python3.5/site-packages/plaidml/keras/backend.py in __call__(self, inputs)
    957             self._invoker.set_output(name, t)
    958 
--> 959         self._invoker.invoke()
    960 
    961         return [t.as_ndarray(_ctx) for t in tensors]

~/test/lib/python3.5/site-packages/plaidml/__init__.py in invoke(self)
   1324 
   1325     def invoke(self):
-> 1326         return Invocation(self._ctx, self)
   1327 
   1328 

~/test/lib/python3.5/site-packages/plaidml/__init__.py in __init__(self, ctx, invoker)
   1330 
   1331     def __init__(self, ctx, invoker):
-> 1332         self._as_parameter_ = _lib().plaidml_schedule_invocation(ctx, invoker)
   1333         self._free = _lib().plaidml_free_invocation
   1334 

~/test/lib/python3.5/site-packages/plaidml/__init__.py in _check_err(self, result, func, args)
    672         if func == self.plaidml_writeback_mapping and args[2]:
    673             return None
--> 674         self.raise_last_status()
    675 
    676 

~/test/lib/python3.5/site-packages/plaidml/library.py in raise_last_status(self)
    134 
    135     def raise_last_status(self):
--> 136         raise self.last_status()
    137 
    138     def _logger_callback(self, unused_arg, level, msg):

Unknown: b'Duplicate binding discovered: X_I_2'
```

I'm on a Macbook Pro with an AMD Radeon R9 M370X and PlaidML 0.1.2. Running the same code on CPU with the Tensorflow background doesn't produce this error.

**Comment by jbruestle** [comment]
Thanks for the detailed report!  I have a repro and I'm looking into it.

**Comment by jbruestle** [comment]
I've figured out the root cause.  I've gotten pulled off on a few other projects but should be able to get a fix out by early next week, possibly sooner.  Sorry for the delay.

**Comment by earhart** [comment]
Thanks again for the bug report!  The fix should be in our next release.

**Comment by jbruestle** [comment]
Reopening until release is officially available.

**Comment by brianretford** [comment]
Should be fixed at head and in 0.1.3
